cmd/argocd-application-controller/commands/argocd_application_controller.go

```diff
@@ -188,10 +192,26 @@ func NewCommand() *cobra.Command {
 				defer closeTracer()
 			}
 
-			go appController.Run(ctx, statusProcessors, operationProcessors)
+			// Graceful shutdown code
+			sigCh := make(chan os.Signal, 1)
+			signal.Notify(sigCh, os.Interrupt, syscall.SIGTERM)
+			wg := sync.WaitGroup{}
+			wg.Add(1)
+			go func() {
+				s := <-sigCh
+				log.Printf("got signal %v, attempting graceful shutdown", s)
+				cancel()
+				wg.Done()
+			}()
+
+			go appController.Run(ctx, statusProcessors, operationProcessors, &wg)
 
 			// Wait forever
-			select {}
+			// select {}
+			wg.Wait()
```

Comment:
Was testing a bit locally, and I think we can get the same effect by just replacing `select {}` with `<-ctx.Done()`. At least the behavior seems the same. What do you think?

Reply:
Wg.Done() looks better because it can be an infinite atmosphere with the existing select{}. And above all, if you consider unity elsewhere, wg.Done() looks better. What do you think?

Reply:
> it can be an infinite atmosphere with the existing select{}

Can you rephrase? I'm not sure what you mean "infinite atmosphere." I'm suggesting removing `select {}` and instead doing `<-ctx.Done()`.

> if you consider unity elsewhere, wg.Done() looks better

If you're looking at the repo-server graceful shutdown code... I wrote that code, and now I don't like it. 😆 Looks like it can be simplified.

Reply:
Deleted select {}. The controller you're working on now doesn't seem to have a problem.

I'll think about repo-server in the future as well. And If possible, I'll also proceed with refactoring on the repo-server side.

Reply:
Can you take a look at this? I think we can simplify further. https://github.com/leehosu/argo-cd/pull/1

Reply:
So Cool. That's simple. I checked and merged into the branch.